hardhat.config.ts

```diff
@@ -0,0 +1,106 @@
+import { config as dotenvConfig } from "dotenv";
+import { resolve } from "path";
+dotenvConfig({ path: resolve(__dirname, "./.env") });
+import "@nomiclabs/hardhat-etherscan";
+import "hardhat-contract-sizer";
+import { HardhatUserConfig } from "hardhat/config";
+import { NetworkUserConfig } from "hardhat/types";
+import "@nomiclabs/hardhat-waffle";
+import "@nomiclabs/hardhat-ganache";
+import "hardhat-typechain";
+import "solidity-coverage";
+import "./tasks/accounts";
+import "./tasks/clean";
+
+const chainIds = {
+  ganache: 1337,
+  goerli: 5,
+  hardhat: 31337,
+  kovan: 42,
+  mainnet: 1,
+  rinkeby: 4,
+  ropsten: 3,
+};
+
+// Ensure that we have all the environment variables we need.
+let mnemonic: string;
+if (!process.env.MNEMONIC) {
+  throw new Error("Please set your MNEMONIC in a .env file");
+} else {
+  mnemonic = process.env.MNEMONIC;
+}
+
+let infuraApiKey: string;
+if (!process.env.INFURA_API_KEY) {
+  throw new Error("Please set your INFURA_API_KEY in a .env file");
+} else {
+  infuraApiKey = process.env.INFURA_API_KEY;
+}
+
+let etherscanApiKey: string;
+if (!process.env.ETHERSCAN_API_KEY) {
+  throw new Error("Please set your ETHERSCAN_API_KEY in a .env file");
+} else {
+  etherscanApiKey = process.env.ETHERSCAN_API_KEY;
+}
+
+let alchemyToken: string;
+if (!process.env.ALCHEMY_TOKEN) {
+  throw new Error("Please set your ALCHEMY_TOKEN in a .env file");
+} else {
+  alchemyToken = process.env.ALCHEMY_TOKEN;
+}
+
+function createTestnetConfig(network: keyof typeof chainIds): NetworkUserConfig {
+  const url: string = "https://" + network + ".infura.io/v3/" + infuraApiKey;
```

Comment:
You don't need to type with `: string` since it's implicitly inferred and string templating is better to embed expressions
```
  const url = `https://${network}.infura.io/v3/${infuraApiKey}`;
```

Reply:
This comes from https://github.com/paulrberg/solidity-template we should push this request upstream if important